<!doctype html>



  


<html class="theme-next mist use-motion">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>



<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />












  
  
  <link href="/vendors/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/vendors/font-awesome/css/font-awesome.min.css?v=4.4.0" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.0.1" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="samtools,software,bioinformatics," />





  <link rel="alternate" href="/atom.xml" title="寂寞先生" type="application/atom+xml" />




  <link rel="shortcut icon" type="image/x-icon" href="http://7xubqr.com1.z0.glb.clouddn.com/favicon.ico?v=5.0.1" />






<meta name="description" content="一、 SAMtools和BCFtools简介SAMtools是一个用于操作sam和bam文件的工具合集，包含有许多命令。BCFtools主要是用来操作vcf和BCF文件的工具合集，包含有许多命令。
二、 SAMtools和BCFtools的安装2.1 对于老版本在0.1.19这个版本及以前版本中samtools和bcftools是整合到一起的，下载链接：https://github.com/sam">
<meta property="og:type" content="article">
<meta property="og:title" content="SAMtools和BCFtools工具详解">
<meta property="og:url" content="http://yoursite.com/2016/05/24/SAMtools和BCFtools工具详解/index.html">
<meta property="og:site_name" content="寂寞先生">
<meta property="og:description" content="一、 SAMtools和BCFtools简介SAMtools是一个用于操作sam和bam文件的工具合集，包含有许多命令。BCFtools主要是用来操作vcf和BCF文件的工具合集，包含有许多命令。
二、 SAMtools和BCFtools的安装2.1 对于老版本在0.1.19这个版本及以前版本中samtools和bcftools是整合到一起的，下载链接：https://github.com/sam">
<meta property="og:updated_time" content="2016-05-24T12:48:04.943Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="SAMtools和BCFtools工具详解">
<meta name="twitter:description" content="一、 SAMtools和BCFtools简介SAMtools是一个用于操作sam和bam文件的工具合集，包含有许多命令。BCFtools主要是用来操作vcf和BCF文件的工具合集，包含有许多命令。
二、 SAMtools和BCFtools的安装2.1 对于老版本在0.1.19这个版本及以前版本中samtools和bcftools是整合到一起的，下载链接：https://github.com/sam">



<script type="text/javascript" id="hexo.configuration">
  var NexT = window.NexT || {};
  var CONFIG = {
    scheme: 'Mist',
    sidebar: {"position":"left","display":"post"},
    fancybox: true,
    motion: true,
    duoshuo: {
      userId: 0,
      author: '博主'
    }
  };
</script>

  <title> SAMtools和BCFtools工具详解 | 寂寞先生 </title>
</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  



  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "//hm.baidu.com/hm.js?58c4dae5f3d2517753d1714e36369277";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>








  
  
    
  

  <div class="container one-collumn sidebar-position-left page-post-detail ">
    <div class="headband"></div>
	
	<!----------- add Fork me on Github ------------>
	
	<a href="https://github.com/starsyi/starsyi.github.com"><img style="position: absolute; top: 0; left: 0; border: 0;" src="https://camo.githubusercontent.com/c6625ac1f3ee0a12250227cf83ce904423abf351/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f6c6566745f677261795f3664366436642e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_left_gray_6d6d6d.png"></a>
	
	<!----------- add Fork me on Github ------------> 

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-meta ">
  

  <div class="custom-logo-site-title">
    <a href="/"  class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <span class="site-title">寂寞先生</span>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>
  <p class="site-subtitle">莫厌追欢笑语频，寻思离乱好伤神。闲来屈指从头数，得见清平有几人</p>
</div>

<div class="site-nav-toggle">
  <button>
    <span class="btn-bar"></span>
    <span class="btn-bar"></span>
    <span class="btn-bar"></span>
  </button>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            归档
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            关于
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="#" class="st-search-show-outputs">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            搜索
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <form class="site-search-form">
  <input type="text" id="st-search-input" class="st-search-input st-default-search-input" />
</form>
<script type="text/javascript">
  (function(w,d,t,u,n,s,e){w['SwiftypeObject']=n;w[n]=w[n]||function(){
  (w[n].q=w[n].q||[]).push(arguments);};s=d.createElement(t);
  e=d.getElementsByTagName(t)[0];s.async=1;s.src=u;e.parentNode.insertBefore(s,e);
  })(window,document,'script','//s.swiftypecdn.com/install/v2/st.js','_st');
  
  _st('install', 'yYtz9jniQj6gr1XndP34','2.0.0');
</script>



    </div>
  
</nav>

 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  
  

  
  
  

  <article class="post post-type-normal " itemscope itemtype="http://schema.org/Article">

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
            
            
              
                SAMtools和BCFtools工具详解
              
            
          </h1>
        

        <div class="post-meta">
          <span class="post-time">
            <span class="post-meta-item-icon">
              <i class="fa fa-calendar-o"></i>
            </span>
            <span class="post-meta-item-text">发表于</span>
            <time itemprop="dateCreated" datetime="2016-05-24T20:43:24+08:00" content="2016-05-24">
              2016-05-24
            </time>
          </span>
		  
		
			<span id="busuanzi_container_page_pv">&nbsp;&nbsp;|&nbsp;&nbsp;阅读量 <span id="busuanzi_value_page_pv"></span> 次</span>
		

          
            <span class="post-category" >
              &nbsp; | &nbsp;
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
              
                <span itemprop="about" itemscope itemtype="https://schema.org/Thing">
                  <a href="/categories/software/" itemprop="url" rel="index">
                    <span itemprop="name">software</span>
                  </a>
                </span>

                
                

              
            </span>
          

          
            
              <span class="post-comments-count">
                &nbsp; | &nbsp;
                <a href="/2016/05/24/SAMtools和BCFtools工具详解/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count ds-thread-count" data-thread-key="2016/05/24/SAMtools和BCFtools工具详解/" itemprop="commentsCount"></span>
                </a>
              </span>
            
          

          

          
          

          
        </div>
      </header>
    


    <div class="post-body" itemprop="articleBody">

      
      

      
        <h2 id="一、-SAMtools和BCFtools简介"><a href="#一、-SAMtools和BCFtools简介" class="headerlink" title="一、 SAMtools和BCFtools简介"></a>一、 SAMtools和BCFtools简介</h2><p>SAMtools是一个用于操作sam和bam文件的工具合集，包含有许多命令。<br>BCFtools主要是用来操作vcf和BCF文件的工具合集，包含有许多命令。</p>
<h2 id="二、-SAMtools和BCFtools的安装"><a href="#二、-SAMtools和BCFtools的安装" class="headerlink" title="二、 SAMtools和BCFtools的安装"></a>二、 SAMtools和BCFtools的安装</h2><h3 id="2-1-对于老版本"><a href="#2-1-对于老版本" class="headerlink" title="2.1 对于老版本"></a>2.1 对于老版本</h3><p>在0.1.19这个版本及以前版本中samtools和bcftools是整合到一起的，下载链接：<a href="https://github.com/samtools/samtools/archive/0.1.19.tar.gz" target="_blank" rel="external">https://github.com/samtools/samtools/archive/0.1.19.tar.gz</a><br>在下载之后解压，直接make，然后将Samtools和bcftools直接添加到环境变量中就可以使用；<br><figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="symbol">$</span> wget https:<span class="comment">//github.com/samtools/samtools/archive/0.1.19.tar.gz</span></span><br><span class="line"><span class="symbol">$</span> tar zxf samtools<span class="number">-0.1</span><span class="number">.19</span>.tar.gz</span><br><span class="line"><span class="symbol">$</span> cd samtools<span class="number">-0.1</span><span class="number">.19</span></span><br><span class="line"><span class="symbol">$</span> make </span><br><span class="line"><span class="symbol">$</span> echo <span class="string">'PATH=$PATH:/your/samtools/path'</span> &gt;&gt; ~/.bashrc</span><br><span class="line"><span class="symbol">$</span> echo <span class="string">'PATH=$PATH:/your/bcftools/path'</span> &gt;&gt; ~/.bashrc</span><br><span class="line"><span class="symbol">$</span> source ~/.bashrc</span><br></pre></td></tr></table></figure></p>
<h3 id="2-2-对于新版本"><a href="#2-2-对于新版本" class="headerlink" title="2.2 对于新版本"></a>2.2 对于新版本</h3><p>对于最新版本的下载：<a href="www.htslib.org">www.htslib.org</a>，SAMtools 新的 1.0 版本改进比较大。 SAMtools 包含 3 部分：<br><figure class="highlight groovy"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">Samtools :</span> 对 SAM<span class="regexp">/BAM/</span>CRAM 格式文件进行读、写、索引和查看操作。</span><br><span class="line"><span class="string">BCFtools :</span> 对 BCF2<span class="regexp">/VCF/</span>gVCF 格式文件进行读写操作； 对 SNP 和 <span class="keyword">short</span> indel 进行 calling<span class="regexp">/filtering/</span>summarising</span><br><span class="line"><span class="string">HTSlib :</span> 用于高通量数据读写操作的 C library</span><br></pre></td></tr></table></figure></p>
<p>老版本(0.1.19)的概要：<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">samtools view -bt ref_list<span class="selector-class">.txt</span> -o aln<span class="selector-class">.bam</span> aln<span class="selector-class">.sam</span><span class="selector-class">.gz</span></span><br><span class="line">samtools sort aln<span class="selector-class">.bam</span> aln<span class="selector-class">.sorted</span></span><br><span class="line">samtools index aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools idxstats aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools view aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> chr2:<span class="number">20</span>,<span class="number">100</span>,<span class="number">000</span>-<span class="number">20</span>,<span class="number">200</span>,<span class="number">000</span></span><br><span class="line">samtools merge out<span class="selector-class">.bam</span> in1<span class="selector-class">.bam</span> in2<span class="selector-class">.bam</span> in3<span class="selector-class">.bam</span></span><br><span class="line">samtools faidx ref<span class="selector-class">.fasta</span></span><br><span class="line">samtools pileup -vcf ref<span class="selector-class">.fasta</span> aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools mpileup -C50 -gf ref<span class="selector-class">.fasta</span> -r chr3:<span class="number">1</span>,<span class="number">000</span>-<span class="number">2</span>,<span class="number">000</span> in1<span class="selector-class">.bam</span> in2<span class="selector-class">.bam</span></span><br><span class="line">samtools tview aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> ref<span class="selector-class">.fasta</span></span><br><span class="line">bcftools index <span class="keyword">in</span><span class="selector-class">.bcf</span></span><br><span class="line">bcftools view <span class="keyword">in</span><span class="selector-class">.bcf</span> chr2:<span class="number">100</span>-<span class="number">200</span> &gt; out<span class="selector-class">.vcf</span></span><br><span class="line">bcftools view -Nvm0.<span class="number">99</span> <span class="keyword">in</span><span class="selector-class">.bcf</span> &gt; out<span class="selector-class">.vcf</span> <span class="number">2</span>&gt; out.afs</span><br></pre></td></tr></table></figure></p>
<p>新版的 SAMtools 将 SAMtools 和 BCFtool 分离开，成为 2 个独立的软件包。 这 2 个都运用了 HTSlib，并且其软件包中都包含了 HTSlib 。<br>最新版的BCFtools和SAMtools也可以通过Github下载：<br><figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="symbol">$</span> git clone --branch=develop git:<span class="comment">//github.com/samtools/htslib.git</span></span><br><span class="line"><span class="symbol">$</span> git clone --branch=develop </span><br><span class="line"><span class="symbol">$</span> git:<span class="comment">//github.com/samtools/bcftools.git</span></span><br><span class="line"><span class="symbol">$</span> git clone --branch=develop </span><br><span class="line"><span class="symbol">$</span> git:<span class="comment">//github.com/samtools/samtools.git</span></span><br><span class="line"><span class="symbol">$</span> cd bcftools &amp;&amp; make</span><br><span class="line"><span class="symbol">$</span> cd ../samtools &amp;&amp; make</span><br></pre></td></tr></table></figure></p>
<p>当然你也可以把它安装到自己想安装的目录下,即在上面两次make的时候，使用一下命令：<br><figure class="highlight dos"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ make prefix=/<span class="built_in">path</span>/to/<span class="built_in">dir</span> install</span><br></pre></td></tr></table></figure></p>
<p>最后再将你你的路径添加到环境变量中：<br><figure class="highlight dos"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">$ <span class="built_in">echo</span> '<span class="built_in">PATH</span>=$<span class="built_in">PATH</span>:/your/samtools/bin/<span class="built_in">path</span>' &gt;&gt; ~/.bashrc</span><br><span class="line">$ <span class="built_in">echo</span> '<span class="built_in">PATH</span>=$<span class="built_in">PATH</span>:/your/bcftools/bin/<span class="built_in">path</span>' &gt;&gt; ~/.bashrc</span><br><span class="line">$ source ~/.bashrc</span><br></pre></td></tr></table></figure></p>
<p>新版本(1.3.1)的概要：<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">samtools view -bt ref_list<span class="selector-class">.txt</span> -o aln<span class="selector-class">.bam</span> aln<span class="selector-class">.sam</span><span class="selector-class">.gz</span></span><br><span class="line">samtools sort -T /tmp/aln<span class="selector-class">.sorted</span> -o aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> aln<span class="selector-class">.bam</span></span><br><span class="line">samtools index aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools idxstats aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools flagstat aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools stats aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools bedcov aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools depth aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span></span><br><span class="line">samtools view aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> chr2:<span class="number">20</span>,<span class="number">100</span>,<span class="number">000</span>-<span class="number">20</span>,<span class="number">200</span>,<span class="number">000</span></span><br><span class="line">samtools merge out<span class="selector-class">.bam</span> in1<span class="selector-class">.bam</span> in2<span class="selector-class">.bam</span> in3<span class="selector-class">.bam</span></span><br><span class="line">samtools faidx ref<span class="selector-class">.fasta</span></span><br><span class="line">samtools tview aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> ref<span class="selector-class">.fasta</span></span><br><span class="line">samtools split merged<span class="selector-class">.bam</span></span><br><span class="line">samtools quickcheck in1<span class="selector-class">.bam</span> in2<span class="selector-class">.cram</span></span><br><span class="line">samtools dict -<span class="selector-tag">a</span> GRCh38 -s <span class="string">"Homo sapiens"</span> ref<span class="selector-class">.fasta</span></span><br><span class="line">samtools fixmate <span class="keyword">in</span><span class="selector-class">.namesorted</span><span class="selector-class">.sam</span> out<span class="selector-class">.bam</span></span><br><span class="line">samtools mpileup -C50 -gf ref<span class="selector-class">.fasta</span> -r chr3:<span class="number">1</span>,<span class="number">000</span>-<span class="number">2</span>,<span class="number">000</span> in1<span class="selector-class">.bam</span> in2<span class="selector-class">.bam</span></span><br><span class="line">samtools flags PAIRED,UNMAP,MUNMAP</span><br><span class="line">samtools fastq <span class="selector-tag">input</span><span class="selector-class">.bam</span> &gt; output<span class="selector-class">.fastq</span></span><br><span class="line">samtools fasta <span class="selector-tag">input</span><span class="selector-class">.bam</span> &gt; output<span class="selector-class">.fasta</span></span><br><span class="line">samtools addreplacerg -r <span class="string">'ID:fish'</span> -r <span class="string">'LB:1334'</span> -r <span class="string">'SM:alpha'</span> -o output<span class="selector-class">.bam</span> <span class="selector-tag">input</span><span class="selector-class">.bam</span></span><br><span class="line">samtools collate aln<span class="selector-class">.sorted</span><span class="selector-class">.bam</span> aln<span class="selector-class">.name_collated</span><span class="selector-class">.bam</span></span><br><span class="line">samtools depad <span class="selector-tag">input</span>.bam</span><br></pre></td></tr></table></figure></p>
<h2 id="三、-SAMtools参数详解"><a href="#三、-SAMtools参数详解" class="headerlink" title="三、 SAMtools参数详解"></a>三、 SAMtools参数详解</h2><p>以下主要是针对samtools-0.1.19的说明文档：<a href="http://www.htslib.org/doc/samtools-0.1.19.html" target="_blank" rel="external">http://www.htslib.org/doc/samtools-0.1.19.html</a></p>
<blockquote>
<p>对于最新版的Samtools， 指1.0之后，有大的改动，可能一下命令介绍有些不适用新版本，所以新版参考：<br>SAMtools：<a href="http://www.htslib.org/doc/samtools.html" target="_blank" rel="external">http://www.htslib.org/doc/samtools.html</a><br>BCFtools：<a href="http://www.htslib.org/doc/bcftools.html" target="_blank" rel="external">http://www.htslib.org/doc/bcftools.html</a></p>
</blockquote>
<h3 id="3-1-view"><a href="#3-1-view" class="headerlink" title="3.1 view"></a>3.1 view</h3><p>view命令的主要功能是：将sam文件转换成bam文件；然后对bam文件进行各种操作，比如数据的排序(不属于本命令的功能)和提取(这些操作是对bam文件进行的，因而当输入为sam文件的时候，不能进行该操作)；最后将排序或提取得到的数据输出为bam或sam（默认的）格式。</p>
<p>bam文件优点：bam文件为二进制文件，占用的磁盘空间比sam文本文件小；利用bam二进制文件的运算速度快。</p>
<p>view命令中，对sam文件头部的输入(-t或-T）和输出(-h)是单独的一些参数来控制的</p>
<figure class="highlight haml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools view [options] &lt;in.bam&gt;|&lt;in.sam&gt; [region1 [...]]</span><br><span class="line">默认情况下不加 region，则是输出所有的 region.</span><br><span class="line"></span><br><span class="line">Options: -b       output BAM</span><br><span class="line">                  默认下输出是 SAM 格式文件，该参数设置输出 BAM 格式</span><br><span class="line">         -<span class="ruby">h       print header <span class="keyword">for</span> the SAM output</span><br><span class="line"></span>                  默认下输出的 sam 格式文件不带 header，该参数设定输出sam文件时带 header 信息</span><br><span class="line">         -<span class="ruby">H       print header only (no alignments)</span><br><span class="line"></span>         -<span class="ruby">S       input is SAM</span><br><span class="line"></span>                  默认下输入是 BAM 文件，若是输入是 SAM 文件，则最好加该参数，否则有时候会报错。</span><br><span class="line">         -<span class="ruby">u       uncompressed BAM output (force -b)</span><br><span class="line"></span>                  该参数的使用需要有-b参数，能节约时间，但是需要更多磁盘空间。</span><br><span class="line">         -<span class="ruby">c       Instead of printing the alignments, only count them <span class="keyword">and</span> print the </span><br><span class="line"></span>                  total number. All filter options, such as ‘-f’, ‘-F’ and ‘-q’ , </span><br><span class="line">                  are taken into account.</span><br><span class="line">         -<span class="ruby"><span class="number">1</span>       fast compression (force -b)</span><br><span class="line"></span>         -<span class="ruby">x       output FLAG <span class="keyword">in</span> HEX (samtools-C specific)</span><br><span class="line"></span>         -<span class="ruby">X       output FLAG <span class="keyword">in</span> string (samtools-C specific)</span><br><span class="line"></span>         -<span class="ruby">c       print only the count of matching records</span><br><span class="line"></span>         -<span class="ruby">L FILE  output alignments overlapping the input BED FILE [null]</span><br><span class="line"></span>         -<span class="ruby">t FILE  list of reference names <span class="keyword">and</span> lengths (force -S) [null]</span><br><span class="line"></span>                  使用一个list文件来作为header的输入</span><br><span class="line">         -<span class="ruby">T FILE  reference sequence file (force -S) [null]</span><br><span class="line"></span>                  使用序列fasta文件作为header的输入</span><br><span class="line">         -<span class="ruby">o FILE  output file name [stdout]</span><br><span class="line"></span>         -<span class="ruby">R FILE  list of read groups to be outputted [null]</span><br><span class="line"></span>         -<span class="ruby">f INT   required flag, <span class="number">0</span> <span class="keyword">for</span> unset [<span class="number">0</span>]</span><br><span class="line"></span>         -<span class="ruby">F INT   filtering flag, <span class="number">0</span> <span class="keyword">for</span> unset [<span class="number">0</span>] </span><br><span class="line"></span>                  Skip alignments with bits present in INT [0]</span><br><span class="line">                  数字4代表该序列没有比对到参考序列上</span><br><span class="line">                  数字8代表该序列的mate序列没有比对到参考序列上</span><br><span class="line">         -<span class="ruby">q INT   minimum mapping quality [<span class="number">0</span>]</span><br><span class="line"></span>         -<span class="ruby">l STR   only output reads <span class="keyword">in</span> library STR [null]</span><br><span class="line"></span>         -<span class="ruby">r STR   only output reads <span class="keyword">in</span> read group STR [null]</span><br><span class="line"></span>         -<span class="ruby">s FLOAT fraction of templates to subsample; integer part as seed [-<span class="number">1</span>]</span><br><span class="line"></span>         -<span class="ruby">?       longer help</span></span><br></pre></td></tr></table></figure>
<p>例子：</p>
<figure class="highlight mipsasm"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">将sam文件转换成<span class="keyword">bam文件</span><br><span class="line"></span>$ samtools view -<span class="keyword">bS </span>abc.sam &gt; abc.<span class="keyword">bam</span><br><span class="line"></span>$ samtools view -<span class="keyword">b </span>-S abc.sam -o abc.<span class="keyword">bam</span><br><span class="line"></span></span><br><span class="line">提取比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -<span class="keyword">bF </span><span class="number">4</span> abc.<span class="keyword">bam </span>&gt; abc.F.<span class="keyword">bam</span><br><span class="line"></span></span><br><span class="line">提取paired reads中两条reads都比对到参考序列上的比对结果，只需要把两个<span class="number">4</span>+<span class="number">8</span>的值<span class="number">12</span>作为过滤参数即可</span><br><span class="line">$ samtools view -<span class="keyword">bF </span><span class="number">12</span> abc.<span class="keyword">bam </span>&gt; abc.F12.<span class="keyword">bam</span><br><span class="line"></span></span><br><span class="line">提取没有比对到参考序列上的比对结果</span><br><span class="line">$ samtools view -<span class="keyword">bf </span><span class="number">4</span> abc.<span class="keyword">bam </span>&gt; abc.f.<span class="keyword">bam</span><br><span class="line"></span></span><br><span class="line">提取<span class="keyword">bam文件中比对到caffold1上的比对结果，并保存到sam文件格式</span><br><span class="line"></span>$ samtools view abc.<span class="keyword">bam </span><span class="keyword">scaffold1 </span>&gt; <span class="keyword">scaffold1.sam</span><br><span class="line"></span></span><br><span class="line">提取<span class="keyword">scaffold1上能比对到30k到100k区域的比对结果</span><br><span class="line"></span>$ samtools view abc.<span class="keyword">bam </span><span class="keyword">scaffold1:30000-100000 </span>$gt<span class="comment">; scaffold1_30k-100k.sam</span></span><br><span class="line"></span><br><span class="line">根据fasta文件，将 header 加入到 sam 或 <span class="keyword">bam </span>文件中</span><br><span class="line">$ samtools view -T genome.fasta -h <span class="keyword">scaffold1.sam </span>&gt; <span class="keyword">scaffold1.h.sam</span></span><br></pre></td></tr></table></figure>
<h3 id="3-2-sort"><a href="#3-2-sort" class="headerlink" title="3.2 sort"></a>3.2 sort</h3><p>sort对bam文件进行排序<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">Usage</span>: <span class="selector-tag">samtools</span> <span class="selector-tag">sort</span> <span class="selector-attr">[-n]</span> <span class="selector-attr">[-m &lt;maxMem&gt;]</span> &lt;<span class="selector-tag">in</span><span class="selector-class">.bam</span>&gt; &lt;<span class="selector-tag">out</span><span class="selector-class">.prefix</span>&gt;  </span><br><span class="line"><span class="selector-tag">-m</span> 参数默认下是 500,000,000 即500<span class="selector-tag">M</span>（不支持<span class="selector-tag">K</span>，<span class="selector-tag">M</span>，<span class="selector-tag">G</span>等缩写）。对于处理大数据时，如果内存够用，则设置大点的值，以节约时间。</span><br><span class="line"><span class="selector-tag">-n</span> 设定排序方式按<span class="selector-tag">short</span> <span class="selector-tag">reads</span>的<span class="selector-tag">ID</span>排序。默认下是按序列在<span class="selector-tag">fasta</span>文件中的顺序（即<span class="selector-tag">header</span>）和序列从左往右的位点排序。</span><br></pre></td></tr></table></figure></p>
<h3 id="3-3-merge"><a href="#3-3-merge" class="headerlink" title="3.3 merge"></a>3.3 merge</h3><p>将2个或2个以上的已经sort了的bam文件融合成一个bam文件。融合后的文件不需要则是已经sort过了的。<br><figure class="highlight sql"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Usage:   samtools <span class="keyword">merge</span> [-nr] [-h inh.sam] &lt;out.bam&gt; &lt;in1.bam&gt; &lt;in2.bam&gt;[...]</span><br><span class="line"></span><br><span class="line">Options: -n       <span class="keyword">sort</span> <span class="keyword">by</span> <span class="keyword">read</span> <span class="keyword">names</span></span><br><span class="line">         -r       attach RG tag (inferred <span class="keyword">from</span> <span class="keyword">file</span> <span class="keyword">names</span>)</span><br><span class="line">         -u       uncompressed BAM <span class="keyword">output</span></span><br><span class="line">         -f       overwrite the <span class="keyword">output</span> BAM <span class="keyword">if</span> exist</span><br><span class="line">         <span class="number">-1</span>       <span class="keyword">compress</span> <span class="keyword">level</span> <span class="number">1</span></span><br><span class="line">         -R <span class="keyword">STR</span>   <span class="keyword">merge</span> <span class="keyword">file</span> <span class="keyword">in</span> the specified region <span class="keyword">STR</span> [all]</span><br><span class="line">         -h <span class="keyword">FILE</span>  copy the header <span class="keyword">in</span> <span class="keyword">FILE</span> <span class="keyword">to</span> &lt;out.bam&gt; [in1.bam]</span><br><span class="line"></span><br><span class="line">Note: Samtools<span class="string">' merge does not reconstruct the @RG dictionary in the header. Users</span><br><span class="line">      must provide the correct header with -h, or uses Picard which properly maintains</span><br><span class="line">      the header dictionary in merging.</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-4-index"><a href="#3-4-index" class="headerlink" title="3.4 index"></a>3.4 index</h3><p>必须对bam文件进行默认情况下的排序后，才能进行index。否则会报错。</p>
<p>建立索引后将产生后缀为.bai的文件，用于快速的随机处理。很多情况下需要有bai文件的存在，特别是显示序列比对情况下。比如samtool的tview命令就需要；gbrowse2显示reads的比对图形的时候也需要。<br><figure class="highlight delphi"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools <span class="keyword">index</span> &lt;<span class="keyword">in</span>.bam&gt; [<span class="keyword">out</span>.<span class="keyword">index</span>]</span><br></pre></td></tr></table></figure></p>
<p>例子：<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">以下两种命令结果一样</span><br><span class="line">$ samtools index abc<span class="selector-class">.sort</span><span class="selector-class">.bam</span></span><br><span class="line">$ samtools index abc<span class="selector-class">.sort</span><span class="selector-class">.bam</span> abc<span class="selector-class">.sort</span><span class="selector-class">.bam</span><span class="selector-class">.bai</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-5-faidx"><a href="#3-5-faidx" class="headerlink" title="3.5 faidx"></a>3.5 faidx</h3><p>对fasta文件建立索引,生成的索引文件以.fai后缀结尾。该命令也能依据索引文件快速提取fasta文件中的某一条（子）序列<br><figure class="highlight http"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">Usage</span>: samtools faidx &lt;in.bam&gt; [ [...]]</span><br><span class="line"></span><br><span class="line"><span class="stylus">对基因组文件建立索引</span><br><span class="line">$ samtools faidx genome<span class="selector-class">.fasta</span></span><br><span class="line">生成了索引文件genome<span class="selector-class">.fasta</span><span class="selector-class">.fai</span>,是一个文本文件，分成了<span class="number">5</span>列。第一列是子序列的名称；</span><br><span class="line">第二列是子序列的长度；个人认为“第三列是序列所在的位置”，因为该数字从上往下逐渐变大，</span><br><span class="line">最后的数字是genome.fasta文件的大小；第<span class="number">4</span>和<span class="number">5</span>列不知是啥意思。于是通过此文件，可以定</span><br><span class="line">位子序列在fasta文件在磁盘上的存放位置，直接快速调出子序列。</span><br><span class="line"></span><br><span class="line">由于有索引文件，可以使用以下命令很快从基因组中提取到fasta格式的子序列</span><br><span class="line">$ samtools faidx genome<span class="selector-class">.fasta</span> scffold_10 &gt; scaffold_10.fasta</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-6-tview"><a href="#3-6-tview" class="headerlink" title="3.6 tview"></a>3.6 tview</h3><p>tview能直观的显示出reads比对基因组的情况，和基因组浏览器有点类似。<br><figure class="highlight http"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">Usage</span>: samtools tview &lt;aln.bam&gt; [ref.fasta]</span><br><span class="line"></span><br><span class="line"><span class="mipsasm">当给出参考基因组的时候，会在第一排显示参考基因组的序列，否则，第一排全用N表示。</span><br><span class="line">按下 g ，则提示输入要到达基因组的某一个位点。例子“<span class="keyword">scaffold_10:1000"表示到达第</span><br><span class="line"></span><span class="number">10</span>号<span class="keyword">scaffold的第1000个碱基位点处。</span><br><span class="line"></span>使用H(左）<span class="keyword">J（上）K（下）L（右）移动显示界面。大写字母移动快，小写字母移动慢。</span><br><span class="line"></span>使用空格建向左快速移动（和 L 类似），使用<span class="keyword">Backspace键向左快速移动（和 </span>H 类似）。</span><br><span class="line">Ctrl+H 向左移动<span class="number">1</span>kb碱基距离； Ctrl+L 向右移动<span class="number">1</span>kb碱基距离</span><br><span class="line">可以用颜色标注比对质量，碱基质量，核苷酸等。<span class="number">30</span>～<span class="number">40</span>的碱基质量或比对质量使用白色表示；</span><br><span class="line"><span class="number">20</span>～<span class="number">30</span>黄色；<span class="number">10</span>～<span class="number">20</span>绿色；<span class="number">0</span>～<span class="number">10</span>蓝色。</span><br><span class="line">使用点号<span class="string">'.'</span>切换显示碱基和点号；使用r切换显示read name等</span><br><span class="line">还有很多其它的使用说明，具体按 ？ 键来查看。</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-7-flagstat"><a href="#3-7-flagstat" class="headerlink" title="3.7 flagstat"></a>3.7 flagstat</h3><p>给出BAM文件的比对结果<br><figure class="highlight http"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">Usage</span>: samtools flagstat &lt;in.bam&gt;</span><br><span class="line"></span><br><span class="line"><span class="elixir"><span class="variable">$ </span>samtools flagstat example.bam</span><br><span class="line"><span class="number">11945742</span> + <span class="number">0</span> <span class="keyword">in</span> total (QC-passed reads + QC-failed reads)</span><br><span class="line"><span class="comment">#总共的reads数</span></span><br><span class="line"><span class="number">0</span> + <span class="number">0</span> duplicates</span><br><span class="line"><span class="number">7536364</span> + <span class="number">0</span> mapped (<span class="number">63.09</span>%<span class="symbol">:-nan%</span>)</span><br><span class="line"><span class="comment">#总体上reads的匹配率</span></span><br><span class="line"><span class="number">11945742</span> + <span class="number">0</span> paired <span class="keyword">in</span> sequencing</span><br><span class="line"><span class="comment">#有多少reads是属于paired reads</span></span><br><span class="line"><span class="number">5972871</span> + <span class="number">0</span> read1</span><br><span class="line"><span class="comment">#reads1中的reads数</span></span><br><span class="line"><span class="number">5972871</span> + <span class="number">0</span> read2</span><br><span class="line"><span class="comment">#reads2中的reads数</span></span><br><span class="line"><span class="number">6412042</span> + <span class="number">0</span> properly paired (<span class="number">53.68</span>%<span class="symbol">:-nan%</span>)</span><br><span class="line"><span class="comment">#完美匹配的reads数：比对到同一条参考序列，并且两条reads之间的距离符合设置的阈值</span></span><br><span class="line"><span class="number">6899708</span> + <span class="number">0</span> with itself <span class="keyword">and</span> mate mapped</span><br><span class="line"><span class="comment">#paired reads中两条都比对到参考序列上的reads数</span></span><br><span class="line"><span class="number">636656</span> + <span class="number">0</span> singletons (<span class="number">5.33</span>%<span class="symbol">:-nan%</span>)</span><br><span class="line"><span class="comment">#单独一条匹配到参考序列上的reads数，和上一个相加，则是总的匹配上的reads数。</span></span><br><span class="line"><span class="number">469868</span> + <span class="number">0</span> with mate mapped to a different chr</span><br><span class="line"><span class="comment">#paired reads中两条分别比对到两条不同的参考序列的reads数</span></span><br><span class="line"><span class="number">243047</span> + <span class="number">0</span> with mate mapped to a different chr (mapQ&gt;=<span class="number">5</span>)</span></span><br></pre></td></tr></table></figure></p>
<p>同上一个，只是其中比对质量&gt;=5的reads的数量</p>
<h3 id="3-8-depth"><a href="#3-8-depth" class="headerlink" title="3.8 depth"></a>3.8 depth</h3><p>得到每个碱基位点的测序深度,并输出到标准输出。<br><figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">Usage</span>: <span class="selector-tag">bam2depth</span> <span class="selector-attr">[-r reg]</span> <span class="selector-attr">[-q baseQthres]</span> <span class="selector-attr">[-Q mapQthres]</span> <span class="selector-attr">[-b in.bed]</span> &lt;<span class="selector-tag">in1</span><span class="selector-class">.bam</span>&gt; <span class="selector-attr">[...]</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-9-其他有用的命令"><a href="#3-9-其他有用的命令" class="headerlink" title="3.9 其他有用的命令"></a>3.9 其他有用的命令</h3><p>reheader 替换bam文件的头<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools reheader &lt;<span class="keyword">in</span><span class="selector-class">.header</span><span class="selector-class">.sam</span>&gt; &lt;<span class="keyword">in</span>.bam&gt;</span><br></pre></td></tr></table></figure></p>
<p>cat 连接多个bam文件，适用于非sorted的bam文件<br><figure class="highlight accesslog"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools cat <span class="string">[-h header.sam]</span> <span class="string">[-o out.bam]</span> &lt;in1.bam&gt; &lt;in2.bam&gt; <span class="string">[ ... ]</span></span><br></pre></td></tr></table></figure></p>
<p>idxstats 统计一个表格，4列，分别为”序列名，序列长度，比对上的reads数，unmapped reads number”。第4列应该是paired reads中有一端能匹配到该scaffold上，而另外一端不匹配到任何scaffolds上的reads数。<br><figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats <span class="tag">&lt;<span class="name">aln.bam</span>&gt;</span></span><br></pre></td></tr></table></figure></p>
<h3 id="3-10-将bam文件转换为fastq文件"><a href="#3-10-将bam文件转换为fastq文件" class="headerlink" title="3.10 将bam文件转换为fastq文件"></a>3.10 将bam文件转换为fastq文件</h3><p>有时候，我们需要提取出比对到一段参考序列的reads，进行小范围的分析，以利于debug等。这时需要将bam或sam文件转换为fastq格式。<br>该网站提供了一个bam转换为fastq的程序：<a href="http://www.hudsonalpha.org/gsl/information/software/bam2fastq" target="_blank" rel="external">http://www.hudsonalpha.org/gsl/information/software/bam2fastq</a><br><figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="symbol">$</span> wget http:<span class="comment">//www.hudsonalpha.org/gsl/static/software/bam2fastq-1.1.0.tgz</span></span><br><span class="line"><span class="symbol">$</span> tar zxf bam2fastq<span class="number">-1.1</span><span class="number">.0</span>.tgz</span><br><span class="line"><span class="symbol">$</span> cd bam2fastq<span class="number">-1.1</span><span class="number">.0</span></span><br><span class="line"><span class="symbol">$</span> make</span><br><span class="line"><span class="symbol">$</span> ./bam2fastq &lt;in.bam&gt;</span><br></pre></td></tr></table></figure></p>
<h3 id="3-11-mpileup"><a href="#3-11-mpileup" class="headerlink" title="3.11 mpileup"></a>3.11 mpileup</h3><p>samtools还有个非常重要的命令mpileup，以前为pileup。该命令用于生成bcf文件，再使用bcftools进行SNP和Indel的分析。bcftools是samtool中附带的软件，在samtools的安装文件夹中可以找到。</p>
<p>最常用的参数有2： -f 来输入有索引文件的fasta参考序列； -g 输出到bcf格式。用法和最简单的例子如下<br><figure class="highlight inform7"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Usage: samtools mpileup <span class="comment">[-EBug]</span> <span class="comment">[-C capQcoef]</span> <span class="comment">[-r reg]</span> <span class="comment">[-f in.fa]</span> <span class="comment">[-l list]</span> <span class="comment">[-M capMapQ]</span> <span class="comment">[-Q minBaseQ]</span> <span class="comment">[-q minMapQ]</span> in.bam <span class="comment">[in2.bam <span class="comment">[...]</span>]</span></span><br><span class="line"></span><br><span class="line">$ samtools mpileup -f genome.fasta abc.bam &gt; abc.txt</span><br><span class="line">$ samtools mpileup -gSDf genome.fasta abc.bam &gt; abc.bcf</span><br><span class="line">$ samtools mpileup -guSDf genome.fasta abc.bam | \</span><br><span class="line">           bcftools view -cvNg - &gt; abc.vcf</span><br></pre></td></tr></table></figure></p>
<p>mpileup不使用-u或-g参数时，则不生成二进制的bcf文件，而生成一个文本文件(输出到标准输出)。该文本文件统计了参考序列中每个碱基位点的比对情况；该文件每一行代表了参考序列中某一个碱基位点的比对结果。比如：<br><figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2841</span>    A       <span class="number">11</span>      ,,,...,....     BHIGDGIJ?FF</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2842</span>    C       <span class="number">12</span>      ,<span class="symbol">$</span>,,...,....^I. CFGEGEGGCFF+</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2843</span>    G       <span class="number">11</span>      ,,...,.....     FDDDDCD?DD+</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2844</span>    G       <span class="number">11</span>      ,,...,.....     FA?AAAA&lt;AA+</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2845</span>    G       <span class="number">11</span>      ,,...,.....     F656666166*</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2846</span>    A       <span class="number">11</span>      ,,...,.....     (<span class="number">1.1111</span>)<span class="number">11</span>*</span><br><span class="line">scaffold_1      <span class="number">2847</span>    A       <span class="number">11</span>      ,,+<span class="number">9</span>acggtgaag.+<span class="number">9</span>ACGGTGAAT.+<span class="number">9</span>ACGGTGAAG.+<span class="number">9</span>ACGGTGAAG,+<span class="number">9</span>acggtgaag.+<span class="number">9</span>ACGGTGAAG.+<span class="number">9</span>ACGGTGAAG.+<span class="number">9</span>ACGGTGAAG.+<span class="number">9</span>ACGGTGAAG.+<span class="number">9</span>ACGGTGAAG       %.+....-..)</span><br><span class="line">scaffold_1      <span class="number">2848</span>    N       <span class="number">11</span>      agGGGgGGGGG     !!<span class="symbol">$</span>!!!!!!!!</span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2849</span>    A       <span class="number">11</span>      c<span class="symbol">$</span>,...,.....    !<span class="number">0000000000</span></span><br><span class="line"><span class="function"><span class="title">scaffold_1</span></span>      <span class="number">2850</span>    A       <span class="number">10</span>      ,...,.....      <span class="number">353333333</span></span><br></pre></td></tr></table></figure></p>
<p>mpileup生成的结果包含6行：参考序列名；位置；参考碱基；比对上的reads数；比对情况；比对上的碱基的质量。其中第5列比较复杂,解释如下：</p>
<p>1 ‘.’代表与参考序列正链匹配。<br>2 ‘,’代表与参考序列负链匹配。<br>3 ‘ATCGN’代表在正链上的不匹配。<br>4 ‘atcgn’代表在负链上的不匹配。<br>5 ‘*’代表模糊碱基<br>6 ‘^’代表匹配的碱基是一个read的开始；’^’后面紧跟的ascii码减去33代表比对质量；这两个符号修饰的是后面的碱基，其后紧跟的碱基(.,ATCGatcgNn)代表该read的第一个碱基。<br>7 ‘$’代表一个read的结束，该符号修饰的是其前面的碱基。<br>8 正则式’+[0-9]+[ACGTNacgtn]+’代表在该位点后插入的碱基；比如上例中在scaffold_1的2847后插入了9个长度的碱基acggtgaag。表明此处极可能是indel。<br>9 正则式’-[0-9]+[ACGTNacgtn]+’代表在该位点后缺失的碱基；</p>
<p>pileup具体的参数如下：<br><figure class="highlight livecodeserver"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">输入参数</span><br><span class="line"><span class="number">-6</span>       Assume <span class="keyword">the</span> quality is <span class="keyword">in</span> <span class="keyword">the</span> Illumina <span class="number">1.3</span>+ encoding. -A Do <span class="keyword">not</span> skip anomalous <span class="built_in">read</span> pairs <span class="keyword">in</span> variant calling. </span><br><span class="line">-B       Disable probabilistic realignment <span class="keyword">for</span> <span class="keyword">the</span> computation <span class="keyword">of</span> base alignment quality (BAQ). BAQ is <span class="keyword">the</span> Phred-scaled probability <span class="keyword">of</span> <span class="keyword">a</span> <span class="built_in">read</span> base being misaligned. Applying this option greatly helps <span class="built_in">to</span> reduce <span class="literal">false</span> SNPs caused <span class="keyword">by</span> misalignments. </span><br><span class="line">-b FILE  List <span class="keyword">of</span> input BAM <span class="built_in">files</span>, <span class="literal">one</span> <span class="built_in">file</span> per <span class="built_in">line</span> [<span class="literal">null</span>]</span><br><span class="line">-C INT   Coefficient <span class="keyword">for</span> downgrading mapping quality <span class="keyword">for</span> reads containing excessive mismatches. Given <span class="keyword">a</span> <span class="built_in">read</span> <span class="keyword">with</span> <span class="keyword">a</span> phred-scaled probability q <span class="keyword">of</span> being generated <span class="built_in">from</span> <span class="keyword">the</span> mapped position, <span class="keyword">the</span> <span class="built_in">new</span> mapping quality is about <span class="built_in">sqrt</span>((INT-q)/INT)*INT. A <span class="literal">zero</span> <span class="built_in">value</span> disables this functionality; <span class="keyword">if</span> enabled, <span class="keyword">the</span> recommended <span class="built_in">value</span> <span class="keyword">for</span> BWA is <span class="number">50.</span> [<span class="number">0</span>] </span><br><span class="line">-d INT   At <span class="keyword">a</span> position, <span class="built_in">read</span> maximally INT reads per input BAM. [<span class="number">250</span>] </span><br><span class="line">-E       Extended BAQ computation. This option helps sensitivity especially <span class="keyword">for</span> MNPs, but may hurt specificity <span class="keyword">a</span> little bit. </span><br><span class="line">-f FILE  The faidx-indexed reference <span class="built_in">file</span> <span class="keyword">in</span> <span class="keyword">the</span> FASTA <span class="built_in">format</span>. The <span class="built_in">file</span> can be optionally compressed <span class="keyword">by</span> razip. [<span class="literal">null</span>] </span><br><span class="line">-l FILE  BED <span class="keyword">or</span> position list <span class="built_in">file</span> containing <span class="keyword">a</span> list <span class="keyword">of</span> regions <span class="keyword">or</span> sites where pileup <span class="keyword">or</span> BCF should be generated [<span class="literal">null</span>] </span><br><span class="line">-M INT       cap mapping quality <span class="keyword">at</span> INT [<span class="number">60</span>]</span><br><span class="line">-q INT 	Minimum mapping quality <span class="keyword">for</span> <span class="keyword">an</span> alignment <span class="built_in">to</span> be used [<span class="number">0</span>] </span><br><span class="line">-Q INT 	Minimum base quality <span class="keyword">for</span> <span class="keyword">a</span> base <span class="built_in">to</span> be considered [<span class="number">13</span>]</span><br><span class="line">-r STR 	Only generate pileup <span class="keyword">in</span> region STR [all sites] </span><br><span class="line"></span><br><span class="line">输出参数</span><br><span class="line">-D 	Output per-sample <span class="built_in">read</span> depth (<span class="built_in">require</span> -g/-u)</span><br><span class="line">-g 	Compute genotype likelihoods <span class="keyword">and</span> output them <span class="keyword">in</span> <span class="keyword">the</span> binary call <span class="built_in">format</span> (BCF). </span><br><span class="line">-S 	Output per-sample Phred-scaled strand bias P-<span class="built_in">value</span> (<span class="built_in">require</span> -g/-u) </span><br><span class="line">-u 	Similar <span class="built_in">to</span> -g except that <span class="keyword">the</span> output is uncompressed BCF, which is preferred <span class="keyword">for</span> piping. </span><br><span class="line"></span><br><span class="line">Options <span class="keyword">for</span> Genotype Likelihood Computation (<span class="keyword">for</span> -g <span class="keyword">or</span> -u):</span><br><span class="line">-e INT 	Phred-scaled gap extension sequencing error probability. Reducing INT leads <span class="built_in">to</span> longer indels. [<span class="number">20</span>] </span><br><span class="line">-h INT 	Coefficient <span class="keyword">for</span> modeling homopolymer errors. Given <span class="keyword">an</span> l-<span class="keyword">long</span> homopolymer run, <span class="keyword">the</span> sequencing error <span class="keyword">of</span> <span class="keyword">an</span> indel <span class="keyword">of</span> size s is modeled <span class="keyword">as</span> INT*s/l. [<span class="number">100</span>] </span><br><span class="line">-I 	Do <span class="keyword">not</span> perform INDEL calling </span><br><span class="line">-L INT 	Skip INDEL calling <span class="keyword">if</span> <span class="keyword">the</span> <span class="built_in">average</span> per-sample depth is above INT. [<span class="number">250</span>] </span><br><span class="line">-o INT 	Phred-scaled gap <span class="built_in">open</span> sequencing error probability. Reducing INT leads <span class="built_in">to</span> more indel calls. [<span class="number">40</span>] </span><br><span class="line">-P STR 	Comma dilimited list <span class="keyword">of</span> platforms (determined <span class="keyword">by</span> @RG-PL) <span class="built_in">from</span> which indel candidates are obtained. It is recommended <span class="built_in">to</span> collect indel candidates <span class="built_in">from</span> sequencing technologies that have low indel error rate such <span class="keyword">as</span> ILLUMINA. [all]</span><br></pre></td></tr></table></figure></p>
<h2 id="四、-bcftools命令详解"><a href="#四、-bcftools命令详解" class="headerlink" title="四、 bcftools命令详解"></a>四、 bcftools命令详解</h2><p>bcftools和samtools类似，用于处理vcf(variant call format)文件和bcf(binary call format)文件。前者为文本文件，后者为其二进制文件。</p>
<p>bcftools使用简单，最主要的命令是view命令，其次还有index和cat等命令。index和cat命令和samtools中类似。此处主讲使用view命令来进行SNP和Indel calling。该命令的使用方法和例子为：</p>
<figure class="highlight accesslog"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">$ bcftools view <span class="string">[-AbFGNQSucgv]</span> <span class="string">[-D seqDict]</span> <span class="string">[-l listLoci]</span> <span class="string">[-s listSample]</span> </span><br><span class="line">          <span class="string">[-i gapSNPratio]</span> <span class="string">[-t mutRate]</span> <span class="string">[-p varThres]</span> <span class="string">[-P prior]</span> </span><br><span class="line">          <span class="string">[-1 nGroup1]</span> <span class="string">[-d minFrac]</span> <span class="string">[-U nPerm]</span> <span class="string">[-X permThres]</span> </span><br><span class="line">          <span class="string">[-T trioType]</span> in.bcf <span class="string">[region]</span></span><br><span class="line"></span><br><span class="line">$ bcftools view -cvNg abc.bcf &gt; snp_indel.vcf</span><br></pre></td></tr></table></figure>
<p>生成的结果文件为vcf格式，有10列，分别是：1 参考序列名；2 varianti所在的left-most位置；3 variant的ID（默认未设置，用’.’表示)；4 参考序列的allele；5 variant的allele(有多个alleles，则用’,’分隔);6 variant/reference QUALity;7 FILTers applied;8 variant的信息，使用分号隔开；9 FORMAT of the genotype fields, separated by colon (optional)； 10 SAMPLE genotypes and per-sample information (optional)。<br>例如：<br><figure class="highlight groovy"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">scaffold_1      <span class="number">2847</span>    .       A       AACGGTGAAG      <span class="number">194</span>     .       INDEL;DP=<span class="number">11</span>;VDB=<span class="number">0.0401</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">8</span>,<span class="number">3</span>;MQ=<span class="number">35</span>;FQ=<span class="number">-67.5</span>   <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">235</span>,<span class="number">33</span>,<span class="number">0</span>:<span class="number">63</span></span><br><span class="line">scaffold_1      <span class="number">3908</span>    .       G       A       <span class="number">111</span>     .       DP=<span class="number">13</span>;VDB=<span class="number">0.0085</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">5</span>,<span class="number">7</span>;MQ=<span class="number">42</span>;FQ=<span class="number">-63</span>   <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">144</span>,<span class="number">36</span>,<span class="number">0</span>:<span class="number">69</span></span><br><span class="line">scaffold_1      <span class="number">4500</span>    .       A       G       <span class="number">31.5</span>    .       DP=<span class="number">8</span>;VDB=<span class="number">0.0034</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">3</span>;MQ=<span class="number">42</span>;FQ=<span class="number">-39</span>    <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">64</span>,<span class="number">12</span>,<span class="number">0</span>:<span class="number">21</span></span><br><span class="line">scaffold_1      <span class="number">4581</span>    .       TGGNGG  TGG     <span class="number">145</span>     .       INDEL;DP=<span class="number">8</span>;VDB=<span class="number">0.0308</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">0</span>,<span class="number">8</span>;MQ=<span class="number">42</span>;FQ=<span class="number">-58.5</span>    <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">186</span>,<span class="number">24</span>,<span class="number">0</span>:<span class="number">45</span></span><br><span class="line">scaffold_1      <span class="number">4644</span>    .       G       A       <span class="number">195</span>     .       DP=<span class="number">21</span>;VDB=<span class="number">0.0198</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">10</span>,<span class="number">10</span>;MQ=<span class="number">42</span>;FQ=<span class="number">-87</span> <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">228</span>,<span class="number">60</span>,<span class="number">0</span>:<span class="number">99</span></span><br><span class="line">scaffold_1      <span class="number">4827</span>    .       NACAAAGA        NA      <span class="number">4.42</span>    .       INDEL;DP=<span class="number">1</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">1</span>,<span class="number">0</span>;MQ=<span class="number">40</span>;FQ=<span class="number">-37.5</span>       <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">0</span>/<span class="number">1</span>:<span class="number">40</span>,<span class="number">3</span>,<span class="number">0</span>:<span class="number">3</span></span><br><span class="line">scaffold_1      <span class="number">4854</span>    .       A       G       <span class="number">48</span>      .       DP=<span class="number">6</span>;VDB=<span class="number">0.0085</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">2</span>,<span class="number">1</span>;MQ=<span class="number">41</span>;FQ=<span class="number">-36</span>    <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">80</span>,<span class="number">9</span>,<span class="number">0</span>:<span class="number">16</span></span><br><span class="line">scaffold_1      <span class="number">5120</span>    .       A       G       <span class="number">85</span>      .       DP=<span class="number">8</span>;VDB=<span class="number">0.0355</span>;AF1=<span class="number">1</span>;AC1=<span class="number">2</span>;DP4=<span class="number">0</span>,<span class="number">0</span>,<span class="number">5</span>,<span class="number">3</span>;MQ=<span class="number">42</span>;FQ=<span class="number">-51</span>    <span class="string">GT:</span><span class="string">PL:</span>GQ        <span class="number">1</span>/<span class="number">1</span>:<span class="number">118</span>,<span class="number">24</span>,<span class="number">0</span>:<span class="number">45</span></span><br></pre></td></tr></table></figure></p>
<p>第8列中显示了对variants的信息描述，比较重要，其中的 Tag 的描述如下：<br><figure class="highlight livecodeserver"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">Tag	Format	Description</span><br><span class="line">AF1	double	Max-likelihood estimate <span class="keyword">of</span> <span class="keyword">the</span> site allele frequency (AF) <span class="keyword">of</span> <span class="keyword">the</span> <span class="keyword">first</span> ALT allele</span><br><span class="line">DP	int	Raw <span class="built_in">read</span> depth (<span class="keyword">without</span> quality filtering)</span><br><span class="line">DP4	int[<span class="number">4</span>]	<span class="comment"># high-quality reference forward bases, ref reverse, alternate for and alt rev bases</span></span><br><span class="line">FQ	int	Consensus quality. Positive: sample genotypes different; negative: otherwise</span><br><span class="line">MQ	int	Root-Mean-Square mapping quality <span class="keyword">of</span> covering reads</span><br><span class="line">PC2	int[<span class="number">2</span>]	Phred probability <span class="keyword">of</span> AF <span class="keyword">in</span> group1 samples being larger (,smaller) than <span class="keyword">in</span> group2</span><br><span class="line">PCHI2	double	Posterior weighted chi^<span class="number">2</span> P-<span class="built_in">value</span> between group1 <span class="keyword">and</span> group2 samples</span><br><span class="line">PV4	double[<span class="number">4</span>]	P-<span class="built_in">value</span> <span class="keyword">for</span> strand bias, baseQ bias, mapQ bias <span class="keyword">and</span> tail distance bias</span><br><span class="line">QCHI2	int	Phred-scaled PCHI2</span><br><span class="line">RP	int	<span class="comment"># permutations yielding a smaller PCHI2</span></span><br><span class="line">CLR	int	Phred <span class="built_in">log</span> ratio <span class="keyword">of</span> genotype likelihoods <span class="keyword">with</span> <span class="keyword">and</span> <span class="keyword">without</span> <span class="keyword">the</span> trio/pair constraint</span><br><span class="line">UGT	<span class="keyword">string</span>	Most probable genotype configuration <span class="keyword">without</span> <span class="keyword">the</span> trio constraint</span><br><span class="line">CGT	<span class="keyword">string</span>	Most probable configuration <span class="keyword">with</span> <span class="keyword">the</span> trio constraint</span><br></pre></td></tr></table></figure></p>
<p>bcftools view 的具体参数如下：</p>
<figure class="highlight livecodeserver"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">Input/Output Options:</span><br><span class="line">-A 	Retain all possible alternate alleles <span class="keyword">at</span> variant sites. By default, <span class="keyword">the</span> view <span class="keyword">command</span> <span class="title">discards</span> <span class="title">unlikely</span> <span class="title">alleles</span>.</span><br><span class="line">-b 	Output <span class="keyword">in</span> <span class="keyword">the</span> BCF <span class="built_in">format</span>. The default is VCF.</span><br><span class="line">-D FILE Sequence dictionary (list <span class="keyword">of</span> chromosome names) <span class="keyword">for</span> VCF-&gt;BCF conversion [<span class="literal">null</span>]</span><br><span class="line">-F 	Indicate PL is generated <span class="keyword">by</span> r921 <span class="keyword">or</span> <span class="keyword">before</span> (ordering is different).</span><br><span class="line">-G 	Suppress all individual genotype information.</span><br><span class="line">-l FILE List <span class="keyword">of</span> sites <span class="keyword">at</span> which information are outputted [all sites]</span><br><span class="line">-N 	Skip sites where <span class="keyword">the</span> REF field is <span class="keyword">not</span> A/C/G/T</span><br><span class="line">-Q 	Output <span class="keyword">the</span> QCALL likelihood <span class="built_in">format</span></span><br><span class="line">-s FILE List <span class="keyword">of</span> samples <span class="built_in">to</span> use. The <span class="keyword">first</span> column <span class="keyword">in</span> <span class="keyword">the</span> input gives <span class="keyword">the</span> sample names <span class="keyword">and</span> <span class="keyword">the</span> <span class="keyword">second</span> gives <span class="keyword">the</span> ploidy, which can only be <span class="number">1</span> <span class="keyword">or</span> <span class="number">2.</span> When <span class="keyword">the</span> <span class="number">2</span>nd column is absent, <span class="keyword">the</span> sample ploidy is assumed <span class="built_in">to</span> be <span class="number">2.</span> In <span class="keyword">the</span> output, <span class="keyword">the</span> ordering <span class="keyword">of</span> samples will be identical <span class="built_in">to</span> <span class="keyword">the</span> <span class="literal">one</span> <span class="keyword">in</span> FILE. [<span class="literal">null</span>]</span><br><span class="line">-S 	The input is VCF instead <span class="keyword">of</span> BCF.</span><br><span class="line">-u 	Uncompressed BCF output (force -b). </span><br><span class="line"></span><br><span class="line">Consensus/Variant Calling Options:</span><br><span class="line">-c 	Call variants <span class="keyword">using</span> Bayesian inference. This option automatically invokes option -e.</span><br><span class="line">-d FLOAT When -v is <span class="keyword">in</span> use, skip loci where <span class="keyword">the</span> fraction <span class="keyword">of</span> samples covered <span class="keyword">by</span> reads is below FLOAT. [<span class="number">0</span>]</span><br><span class="line">        当有多个sample用于variants calling时，比如多个转录组数据或多个重测序</span><br><span class="line">        数据需要比对到参考基因组上，设置该值，表明至少有该&lt;float <span class="number">0</span>~<span class="number">1</span>&gt;比例的</span><br><span class="line">        samples在该位点都有覆盖才计算入variant.所以对于只有一个sample的情况</span><br><span class="line">        下，该值设置在<span class="number">0</span>～<span class="number">1</span>之间没有意义，大于<span class="number">1</span>则得不到任何结果。</span><br><span class="line">-e 	Perform <span class="built_in">max</span>-likelihood inference only, including estimating <span class="keyword">the</span> site allele frequency, testing Hardy-Weinberg equlibrium <span class="keyword">and</span> testing associations <span class="keyword">with</span> LRT.</span><br><span class="line">-g 	Call per-sample genotypes <span class="keyword">at</span> variant sites (force -c)</span><br><span class="line">-i FLOAT Ratio <span class="keyword">of</span> INDEL-<span class="built_in">to</span>-SNP mutation rate [<span class="number">0.15</span>]</span><br><span class="line">-p FLOAT A site is considered <span class="built_in">to</span> be <span class="keyword">a</span> variant <span class="keyword">if</span> P(ref|D)</span><br><span class="line">-t FLOAT Scaled muttion rate <span class="keyword">for</span> variant calling [<span class="number">0.001</span>]</span><br><span class="line">-T STR 	Enable pair/trio calling. For trio calling, option -s is usually needed <span class="built_in">to</span> be applied <span class="built_in">to</span> configure <span class="keyword">the</span> trio members <span class="keyword">and</span> their ordering. In <span class="keyword">the</span> <span class="built_in">file</span> supplied <span class="built_in">to</span> <span class="keyword">the</span> option -s, <span class="keyword">the</span> <span class="keyword">first</span> sample must be <span class="keyword">the</span> child, <span class="keyword">the</span> <span class="keyword">second</span> <span class="keyword">the</span> father <span class="keyword">and</span> <span class="keyword">the</span> <span class="keyword">third</span> <span class="keyword">the</span> mother. The valid values <span class="keyword">of</span> STR are ‘pair’, ‘trioauto’, ‘trioxd’ <span class="keyword">and</span> ‘trioxs’, where ‘pair’ calls differences between <span class="literal">two</span> input samples, <span class="keyword">and</span> ‘trioxd’ (‘trioxs’) specifies that <span class="keyword">the</span> input is <span class="built_in">from</span> <span class="keyword">the</span> X chromosome non-PAR regions <span class="keyword">and</span> <span class="keyword">the</span> child is <span class="keyword">a</span> female (male). [<span class="literal">null</span>]</span><br><span class="line">-v 	Output variant sites only (force -c) </span><br><span class="line"></span><br><span class="line">Contrast Calling <span class="keyword">and</span> Association Test Options:</span><br><span class="line"><span class="number">-1</span> INT 	Number <span class="keyword">of</span> group<span class="number">-1</span> samples. This option is used <span class="keyword">for</span> dividing <span class="keyword">the</span> samples <span class="keyword">into</span> <span class="literal">two</span> groups <span class="keyword">for</span> contrast SNP calling <span class="keyword">or</span> association test. When this option is <span class="keyword">in</span> use, <span class="keyword">the</span> following VCF INFO will be outputted: PC2, PCHI2 <span class="keyword">and</span> QCHI2. [<span class="number">0</span>]</span><br><span class="line">-U INT 	Number <span class="keyword">of</span> permutations <span class="keyword">for</span> association test (<span class="keyword">effective</span> only <span class="keyword">with</span> <span class="number">-1</span>) [<span class="number">0</span>]</span><br><span class="line">-X FLOAT Only perform permutations <span class="keyword">for</span> P(chi^<span class="number">2</span>)</span><br></pre></td></tr></table></figure>
<p>使用bcftools得到variant calling结果后。需要对结果再次进行过滤。主要依据比对结果中第8列信息。其中的 DP4 一行尤为重要，提供了4个数据：1 比对结果和正链一致的reads数、2 比对结果和负链一致的reads数、3 比对结果在正链的variant上的reads数、4 比对结果在负链的variant上的reads数。可以设定 （value3 + value4）大于某一阈值，才算是variant。比如：<br><figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ perl -ne '<span class="keyword">print</span> <span class="variable">$_</span> <span class="keyword">if</span> /DP4=(\<span class="keyword">d</span>+),(\<span class="keyword">d</span>+),(\<span class="keyword">d</span>+),(\<span class="keyword">d</span>+)/ &amp;&amp; (<span class="variable">$3</span>+<span class="variable">$4</span>)&gt;=10 &amp;&amp; (<span class="variable">$3</span>+<span class="variable">$4</span>)/(<span class="variable">$1</span>+<span class="variable">$2</span>+<span class="variable">$3</span>+<span class="variable">$4</span>)&gt;=0.8' snp_indel.vcf &gt; snp_indel.final.vcf</span><br></pre></td></tr></table></figure></p>
<h3 id="3-13-samtools-rmdup"><a href="#3-13-samtools-rmdup" class="headerlink" title="3.13 samtools rmdup"></a>3.13 samtools rmdup</h3><p>NGS上机测序前需要进行PCR一步，使一个模板扩增出一簇，从而在上机测序的时候表现出为1个点，即一个reads。若一个模板扩增出了多簇，结果得到了多个reads，这些reads的坐标(coordinates)是相近的。在进行了reads比对后需要将这些由PCR duplicates获得的reads去掉，并只保留最高比对质量的read。使用rmdup命令即可完成.<br><figure class="highlight haml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Usage:  samtools rmdup [-sS]  </span><br><span class="line">-<span class="ruby">s 对single-<span class="keyword">end</span> reads。默认情况下，只对paired-<span class="keyword">end</span> reads</span><br><span class="line"></span>-<span class="ruby">S 将Paired-<span class="keyword">end</span> reads作为single-<span class="keyword">end</span> reads处理。</span><br><span class="line"></span></span><br><span class="line">$ samtools input.sorted.bam output.bam</span><br></pre></td></tr></table></figure></p>
<h2 id="五、-SAMtools常用统计命令"><a href="#五、-SAMtools常用统计命令" class="headerlink" title="五、 SAMtools常用统计命令"></a>五、 SAMtools常用统计命令</h2><p>SAMtools不仅仅用来call snp。从samtools的软件名就能看出，是对SAM格式文件进行操作的工作，比如讲sam转成bam格式，index，rmdup等等。samtools结合linux命令比如grep，awk和SAM格式描述的flag，tag，亦是非常非常非常强大，比如根据flag过滤duplicate的reads，根据XA tag过滤multiple hit的reads。本文在此只介绍一下samtools的统计命令，能快速对bam文件进行各种统计。</p>
<h3 id="5-1-idxstats"><a href="#5-1-idxstats" class="headerlink" title="5.1 idxstats"></a>5.1 idxstats</h3><p>检索和打印与输入文件相对应的index file里的统计信息，所以要对输入的bam文件进行index</p>
<figure class="highlight dns"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">$ samtools idxstats in.sam|in.bam|in.cram</span><br><span class="line">#ref    sequence_length mapped_reads    unmapped_reads</span><br><span class="line">chr1    <span class="number">249250621</span>       <span class="number">4998344</span>         <span class="number">1005</span></span><br><span class="line">chr2    <span class="number">243199373</span>       <span class="number">3020248</span>         <span class="number">595</span></span><br><span class="line">chr3    <span class="number">198022430</span>       <span class="number">2418804</span>         <span class="number">449</span></span><br></pre></td></tr></table></figure>
<h3 id="5-2-bedcov"><a href="#5-2-bedcov" class="headerlink" title="5.2 bedcov"></a>5.2 bedcov</h3><p>计算覆盖到每个区域的总碱基数目<br><figure class="highlight dns"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ samtools bedcov region.bed in1.sam|in1.bam|in1.cram[...]</span><br><span class="line">#chr    start-<span class="number">1</span>     end         totalbase</span><br><span class="line">chr1    <span class="number">100000</span>      <span class="number">1000000</span>     <span class="number">1709228</span></span><br><span class="line">chr2    <span class="number">2000000</span>     <span class="number">65885852</span>    <span class="number">64362582</span></span><br></pre></td></tr></table></figure></p>
<h3 id="5-3-depth"><a href="#5-3-depth" class="headerlink" title="5.3 depth"></a>5.3 depth</h3><p>计算每个位点的深度<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">$ samtools depth [options] [<span class="keyword">in</span>1.sam|<span class="keyword">in</span>1.bam|<span class="keyword">in</span>1.cram [<span class="keyword">in</span>2.sam|<span class="keyword">in</span>2.bam|<span class="keyword">in</span>2.cram] [...]]</span><br><span class="line"><span class="comment">#chr    pos depth</span></span><br><span class="line">chr1    1   5</span><br><span class="line">chr1    2   5</span><br></pre></td></tr></table></figure></p>
<h3 id="5-4-flagstat"><a href="#5-4-flagstat" class="headerlink" title="5.4 flagstat"></a>5.4 flagstat</h3><p>根据flag统计多少map的reads等信息<br><figure class="highlight ruby"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flagstat <span class="keyword">in</span>.sam<span class="params">|<span class="keyword">in</span>.bam|</span><span class="keyword">in</span>.cram</span><br><span class="line"><span class="number">43444444</span> + <span class="number">0</span> <span class="keyword">in</span> total (QC-passed reads + QC-failed reads)</span><br><span class="line"><span class="number">5863846</span> + <span class="number">0</span> secondary</span><br><span class="line"><span class="number">0</span> + <span class="number">0</span> supplementary</span><br><span class="line"><span class="number">0</span> + <span class="number">0</span> duplicates</span><br><span class="line"><span class="number">43431948</span> + <span class="number">0</span> mapped (<span class="number">99.97</span>%<span class="symbol">:-nan%</span>)</span><br><span class="line"><span class="number">37580598</span> + <span class="number">0</span> paired <span class="keyword">in</span> sequencing</span><br></pre></td></tr></table></figure></p>
<h3 id="5-5-stats"><a href="#5-5-stats" class="headerlink" title="5.5 stats"></a>5.5 stats</h3><p>输出的信息比较多，部分如下<br><figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">$ samtools stats [options] <span class="keyword">in</span>.sam|<span class="keyword">in</span>.bam|<span class="keyword">in</span>.cram [region...]</span><br><span class="line">Summary Numbers</span><br><span class="line">raw total sequences，filtered sequences, reads <span class="keyword">mapped</span>, reads <span class="keyword">mapped</span> <span class="keyword">and</span> paired,reads properly paired等信息</span><br><span class="line">Fragment Qualitites</span><br><span class="line">根据cycle统计每个位点上的碱基质量分布</span><br><span class="line">Coverage distribution</span><br><span class="line">深度为<span class="number">1</span>，<span class="number">2</span>，<span class="number">3</span>，，，的碱基数目</span><br><span class="line">ACGT content per cycle</span><br><span class="line">ACGT在每个cycle中的比例</span><br><span class="line">Insert sizes</span><br><span class="line">插入长度的统计</span><br><span class="line"><span class="keyword">Read</span> lengths</span><br><span class="line"><span class="keyword">read</span>的长度分布</span><br></pre></td></tr></table></figure></p>
<p>参考内容：</p>
<ol>
<li><a href="http://www.chenlianfu.com/?p=1399" target="_blank" rel="external">宠辱不惊，一心问学！</a></li>
<li><a href="http://wp.zxzyl.com/?p=132" target="_blank" rel="external">Lablueee’s web site</a></li>
</ol>

      
    </div>

    <div>
      
        
      
    </div>

    <div>
      
        

      
    </div>

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/samtools/" rel="tag">#samtools</a>
          
            <a href="/tags/software/" rel="tag">#software</a>
          
            <a href="/tags/bioinformatics/" rel="tag">#bioinformatics</a>
          
        </div>
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2016/05/24/SAM文件格式说明/" rel="next" title="SAM文件格式说明">
                <i class="fa fa-chevron-left"></i> SAM文件格式说明
              </a>
            
          </div>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </article>



    <div class="post-spread">
      
        <!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <a class="jiathis_button_tsina"></a>
  <a class="jiathis_button_tqq"></a>
  <a class="jiathis_button_weixin"></a>
  <a class="jiathis_button_cqq"></a>
  <a class="jiathis_button_douban"></a>
  <a class="jiathis_button_renren"></a>
  <a class="jiathis_button_qzone"></a>
  <a class="jiathis_button_kaixin001"></a>
  <a class="jiathis_button_copy"></a>
  <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank"></a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" >
  var jiathis_config={
    hideMore:false
  }
</script>
<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js" charset="utf-8"></script>
<!-- JiaThis Button END -->

      
    </div>
  </div>


          </div>
          

  <p>热评文章</p>
  <div class="ds-top-threads" data-range="weekly" data-num-items="4"></div>


          
  <div class="comments" id="comments">
    
      <div class="ds-thread" data-thread-key="2016/05/24/SAMtools和BCFtools工具详解/"
           data-title="SAMtools和BCFtools工具详解" data-url="http://yoursite.com/2016/05/24/SAMtools和BCFtools工具详解/">
      </div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel ">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="http://7xubqr.com1.z0.glb.clouddn.com/avatar.png"
               alt="starsyi" />
          <p class="site-author-name" itemprop="name">starsyi</p>
          <p class="site-description motion-element" itemprop="description">自古求真皆寂寞，唯挑心灯伴夜霭</p>
        </div>
        <nav class="site-state motion-element">
          <div class="site-state-item site-state-posts">
            <a href="/archives">
              <span class="site-state-item-count">5</span>
              <span class="site-state-item-name">日志</span>
            </a>
          </div>

          
            <div class="site-state-item site-state-categories">
              <a href="/categories">
                <span class="site-state-item-count">1</span>
                <span class="site-state-item-name">分类</span>
              </a>
            </div>
          

          
            <div class="site-state-item site-state-tags">
              <a href="/tags">
                <span class="site-state-item-count">6</span>
                <span class="site-state-item-name">标签</span>
              </a>
            </div>
          

        </nav>

        
          <div class="feed-link motion-element">
            <a href="/atom.xml" rel="alternate">
              <i class="fa fa-rss"></i>
              RSS
            </a>
          </div>
        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="https://github.com/starsyi" target="_blank" title="Github">
                  
                    <i class="fa fa-fw fa-globe"></i>
                  
                  Github
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://weibo.com/u/2436657131" target="_blank" title="Weibo">
                  
                    <i class="fa fa-fw fa-weibo"></i>
                  
                  Weibo
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="mailto:yxx285223595@163.com" target="_blank" title="Email">
                  
                    <i class="fa fa-fw fa-globe"></i>
                  
                  Email
                </a>
              </span>
            
          
        </div>

        
        

        
        

      </section>

      
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">
            
              
            
            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#一、-SAMtools和BCFtools简介"><span class="nav-number">1.</span> <span class="nav-text">一、 SAMtools和BCFtools简介</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#二、-SAMtools和BCFtools的安装"><span class="nav-number">2.</span> <span class="nav-text">二、 SAMtools和BCFtools的安装</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#2-1-对于老版本"><span class="nav-number">2.1.</span> <span class="nav-text">2.1 对于老版本</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-2-对于新版本"><span class="nav-number">2.2.</span> <span class="nav-text">2.2 对于新版本</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#三、-SAMtools参数详解"><span class="nav-number">3.</span> <span class="nav-text">三、 SAMtools参数详解</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#3-1-view"><span class="nav-number">3.1.</span> <span class="nav-text">3.1 view</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-2-sort"><span class="nav-number">3.2.</span> <span class="nav-text">3.2 sort</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-3-merge"><span class="nav-number">3.3.</span> <span class="nav-text">3.3 merge</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-4-index"><span class="nav-number">3.4.</span> <span class="nav-text">3.4 index</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-5-faidx"><span class="nav-number">3.5.</span> <span class="nav-text">3.5 faidx</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-6-tview"><span class="nav-number">3.6.</span> <span class="nav-text">3.6 tview</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-7-flagstat"><span class="nav-number">3.7.</span> <span class="nav-text">3.7 flagstat</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-8-depth"><span class="nav-number">3.8.</span> <span class="nav-text">3.8 depth</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-9-其他有用的命令"><span class="nav-number">3.9.</span> <span class="nav-text">3.9 其他有用的命令</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-10-将bam文件转换为fastq文件"><span class="nav-number">3.10.</span> <span class="nav-text">3.10 将bam文件转换为fastq文件</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-11-mpileup"><span class="nav-number">3.11.</span> <span class="nav-text">3.11 mpileup</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#四、-bcftools命令详解"><span class="nav-number">4.</span> <span class="nav-text">四、 bcftools命令详解</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#3-13-samtools-rmdup"><span class="nav-number">4.1.</span> <span class="nav-text">3.13 samtools rmdup</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#五、-SAMtools常用统计命令"><span class="nav-number">5.</span> <span class="nav-text">五、 SAMtools常用统计命令</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#5-1-idxstats"><span class="nav-number">5.1.</span> <span class="nav-text">5.1 idxstats</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-2-bedcov"><span class="nav-number">5.2.</span> <span class="nav-text">5.2 bedcov</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-3-depth"><span class="nav-number">5.3.</span> <span class="nav-text">5.3 depth</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-4-flagstat"><span class="nav-number">5.4.</span> <span class="nav-text">5.4 flagstat</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-5-stats"><span class="nav-number">5.5.</span> <span class="nav-text">5.5 stats</span></a></li></ol></li></ol></div>
            
          </div>
        </section>
      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy;  2016.05 - 
  <span itemprop="copyrightYear">2016</span>
  <span class="with-love">
    <i class="icon-next-heart fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">starsyi

  &nbsp;&nbsp;|&nbsp;&nbsp;
  <script type="text/javascript">
    var cnzz_protocol = (("https:" == document.location.protocol) ? " https://" : " http://");
	document.write(unescape("%3Cspan id='cnzz_stat_icon_1259228303'%3E%3C/span%3E%3Cscript src='" + cnzz_protocol + "s4.cnzz.com/z_stat.php%3Fid%3D1259228303' type='text/javascript'%3E%3C/script%3E"));
	</script>

  &nbsp;&nbsp;|&nbsp;&nbsp;<span><a href="/sitemap.xml">Google网站地图</a></span>
  &nbsp;&nbsp;|&nbsp;&nbsp;<span><a href="/baidusitemap.xml">百度网站地图</a></span>

  </span>
</div>

<div class="powered-by">
  由 <a class="theme-link" href="http://hexo.io">Hexo</a> 强力驱动
</div>

<div class="theme-info">
  主题 -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next">
    NexT.Mist
  </a>
</div>

&nbsp;&nbsp;|&nbsp;&nbsp;本页点击 <span id="busuanzi_value_page_pv"></span> 次
&nbsp;&nbsp;|&nbsp;&nbsp;本站总点击 <span id="busuanzi_value_site_pv"></span> 次
&nbsp;&nbsp;|&nbsp;&nbsp;您是第 <span id="busuanzi_value_site_uv"></span> 位访客

<script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js">
</script>

<script>
(function(){
    var bp = document.createElement('script');
    bp.src = '//push.zhanzhang.baidu.com/push.js';
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


        

        
      </div>
    </footer>

    <div class="back-to-top">
      <i class="fa fa-arrow-up"></i>
    </div>
  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  



  
  <script type="text/javascript" src="/vendors/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/vendors/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/vendors/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/vendors/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/vendors/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/vendors/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.0.1"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.0.1"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.0.1"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.0.1"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.0.1"></script>



  

  
    
  

  <script type="text/javascript">
    var duoshuoQuery = {short_name:"starsyi"};
    (function() {
      var ds = document.createElement('script');
      ds.type = 'text/javascript';ds.async = true;
      ds.id = 'duoshuo-script';
      ds.src = (document.location.protocol == 'https:' ? 'https:' : 'http:') + '/js/embed.js';
      ds.charset = 'UTF-8';
      (document.getElementsByTagName('head')[0]
      || document.getElementsByTagName('body')[0]).appendChild(ds);
    })();
  </script>

  
    
  





  
  
  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length == 0) {
       search_path = "search.xml";
    }
    var path = "/" + search_path;
    // monitor main search box;

    function proceedsearch() {
      $("body").append('<div class="popoverlay">').css('overflow', 'hidden');
      $('.popup').toggle();

    }
    // search function;
    var searchFunc = function(path, search_id, content_id) {
    'use strict';
    $.ajax({
        url: path,
        dataType: "xml",
        async: true,
        success: function( xmlResponse ) {
            // get the contents from search data
            isfetched = true;
            $('.popup').detach().appendTo('.header-inner');
            var datas = $( "entry", xmlResponse ).map(function() {
                return {
                    title: $( "title", this ).text(),
                    content: $("content",this).text(),
                    url: $( "url" , this).text()
                };
            }).get();
            var $input = document.getElementById(search_id);
            var $resultContent = document.getElementById(content_id);
            $input.addEventListener('input', function(){
                var matchcounts = 0;
                var str='<ul class=\"search-result-list\">';                
                var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
                $resultContent.innerHTML = "";
                if (this.value.trim().length > 1) {
                // perform local searching
                datas.forEach(function(data) {
                    var isMatch = true;
                    var content_index = [];
                    var data_title = data.title.trim().toLowerCase();
                    var data_content = data.content.trim().replace(/<[^>]+>/g,"").toLowerCase();
                    var data_url = data.url;
                    var index_title = -1;
                    var index_content = -1;
                    var first_occur = -1;
                    // only match artiles with not empty titles and contents
                    if(data_title != '' && data_content != '') {
                        keywords.forEach(function(keyword, i) {
                            index_title = data_title.indexOf(keyword);
                            index_content = data_content.indexOf(keyword);
                            if( index_title < 0 && index_content < 0 ){
                                isMatch = false;
                            } else {
                                if (index_content < 0) {
                                    index_content = 0;
                                }
                                if (i == 0) {
                                    first_occur = index_content;
                                }
                            }
                        });
                    }
                    // show search results
                    if (isMatch) {
                        matchcounts += 1;
                        str += "<li><a href='"+ data_url +"' class='search-result-title'>"+ data_title +"</a>";
                        var content = data.content.trim().replace(/<[^>]+>/g,"");
                        if (first_occur >= 0) {
                            // cut out 100 characters
                            var start = first_occur - 20;
                            var end = first_occur + 80;
                            if(start < 0){
                                start = 0;
                            }
                            if(start == 0){
                                end = 50;
                            }
                            if(end > content.length){
                                end = content.length;
                            }
                            var match_content = content.substring(start, end);
                            // highlight all keywords
                            keywords.forEach(function(keyword){
                                var regS = new RegExp(keyword, "gi");
                                match_content = match_content.replace(regS, "<b class=\"search-keyword\">"+keyword+"</b>");
                            });
                            
                            str += "<p class=\"search-result\">" + match_content +"...</p>"
                        }
                        str += "</li>";
                    }
                })};
                str += "</ul>";
                if (matchcounts == 0) { str = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>' }
                if (keywords == "") { str = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>' }
                $resultContent.innerHTML = str;
            });
            proceedsearch();
        }
    });}

    // handle and trigger popup window;
    $('.popup-trigger').mousedown(function(e) {
      e.stopPropagation();
      if (isfetched == false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };

    });

    $('.popup-btn-close').click(function(e){
      $('.popup').hide();
      $(".popoverlay").remove();
      $('body').css('overflow', '');
    });
    $('.popup').click(function(e){
      e.stopPropagation();
    });
  </script>

  
  <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
        processEscapes: true,
        skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      }
    });
  </script>

  <script type="text/x-mathjax-config">
    MathJax.Hub.Queue(function() {
      var all = MathJax.Hub.getAllJax(), i;
      for (i=0; i < all.length; i += 1) {
        all[i].SourceElement().parentNode.className += ' has-jax';
      }
    });
  </script>
  <script type="text/javascript" src="//cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


  

  

</body>
</html>
